crates/cadenza-eval/src/interner.rs

```diff
@@ -33,6 +56,262 @@ impl fmt::Display for InternedId {
+    }
+}
+
+impl<S: Storage> fmt::Display for Interned<S>
+where
+    S::Value: fmt::Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&**self, f)
+    }
+}
+
+// =============================================================================
+// Macro to define string storage
+// =============================================================================
+
+/// A macro to define string storage for interning.
+///
+/// This creates a ZST marker type and implements `Storage` for it
+/// using thread-local storage. The generated type does not implement
+/// `Sync` or `Send` to prevent the interned values from being sent
+/// to a different thread.
+///
+/// # Example
+///
+/// ```
+/// use cadenza_eval::define_string_storage;
+/// use cadenza_eval::interner::Interned;
+///
+/// define_string_storage!(MyStrings);
+///
+/// let s: Interned<MyStrings> = Interned::from("hello");
+/// assert_eq!(&*s, "hello");
+/// ```
+#[macro_export]
+macro_rules! define_string_storage {
+    ($name:ident) => {
+        /// A ZST marker for string storage.
+        ///
+        /// This type does not implement `Sync` or `Send` to ensure
+        /// interned values cannot be sent to a different thread.
+        #[derive(Debug, Clone, Copy, Default)]
+        pub struct $name {
+            // PhantomData to prevent Sync/Send
+            _not_send_sync: ::std::marker::PhantomData<*const ()>,
+        }
+
+        #[allow(dead_code)]
+        impl $name {
+            /// Creates a new storage marker.
+            pub const fn new() -> Self {
+                Self {
+                    _not_send_sync: ::std::marker::PhantomData,
+                }
+            }
+        }
+
+        ::std::thread_local! {
+            static STORAGE: ::std::cell::RefCell<$crate::interner::StringStorageData> =
+                ::std::cell::RefCell::new($crate::interner::StringStorageData::new());
+        }
+
+        impl $crate::interner::Storage for $name {
+            type Index = u32;
+            type Value = str;
+
+            fn insert(value: &str) -> Self::Index {
+                STORAGE.with(|data| data.borrow_mut().insert(value))
+            }
+
+            fn resolve(index: Self::Index) -> &'static Self::Value {
+                STORAGE.with(|data| {
+                    // SAFETY: The string lives in thread-local storage for the lifetime
+                    // of the thread. We return a 'static reference because the storage
+                    // outlives any single call. The PhantomData in the storage marker
+                    // prevents Send/Sync, ensuring the reference isn't used across threads.
+                    let borrowed = data.borrow();
+                    let s = borrowed.get(index);
+                    unsafe { ::std::mem::transmute::<&str, &'static str>(s) }
+                })
+            }
+        }
+    };
+}
+
+/// Internal data structure for string storage.
```

Comment:
If this is internal why is it public? That being said, I don't want to overcomplicate things right now. We only know of 3 types we need to intern right now so let's just focus on that. We have string literals mapping to strings. We have integer literals mapping to `Result<i64, IntParseError>`. And we have floats doing the same. Outside of that, we're good for now. Additionally, even though I mentioned local storage, we don't need it. Let's just make a static OnceLock and call it a day.